### ISSUE TITLE
Add `random_init` flag to `setup_finetune.yaml` for architecture-only baselines

### ISSUE THREAD

**Issue by msalganik** [niznik-dev/cruijff_kit#438]
We discussed this feature at our research meeting. The way we implemented in Cruijff does not make sense for Cruijff Kit.  

I worked with Claude to develop this issue.  It seems reasonable, but I'm sure that it could be improved.  Please treat this as a rough draft.

## Why this should be a first-class feature in cruijff_kit

Random-weight baselines are standard in the interpretability and architecture-analysis literature, and the pieces already exist in cruijff_kit — torchtune's builder functions return un-initialized models by design, and cruijff_kit's custom recipes own the checkpoint-loading step. What's missing is a single flag in `setup_finetune.yaml` that says "skip the checkpoint step." The workaround below works, but it requires `tune cp`-ing a cruijff_kit recipe and hand-editing it, which is non-obvious and undocumented. A `random_init: true` flag turns a half-day of research friction into a one-line config change.

## Summary

Add a `random_init: bool = false` field to `setup_finetune.yaml`. When set, the scaffold renders a `finetune.yaml` that skips checkpoint loading entirely and the recipe trains from the builder's initial weights.

## Proposed API

```yaml
# setup_finetune.yaml
torchtune_model_name: Llama-3.2-1B-Instruct
custom_recipe: cruijff_kit.tools.torchtune.custom_recipes.lora_finetune_single_device_stable
random_init: true
seed: 42
# ... rest of config unchanged (dataset, LoRA, LR, etc.)
```

User workflow is otherwise identical:

```bash
python ../../tools/torchtune/setup_finetune.py
sbatch finetune.slurm
```

## Current workaround

A user wanting a random-init training run today has to:

1. Copy the relevant cruijff_kit custom recipe file (`tools/torchtune/custom_recipes/lora_finetune_single_device_stable.py` or `..._distributed_stable.py`) into their experiment directory
2. Guard the `model.load_state_dict(...)` call by hand
3. Fabricate a dummy checkpoint directory, or also guard the checkpointer instantiation
4. Point `custom_recipe:` in `setup_finetune.yaml` at the edited copy

This works but is fiddly and undocumented. The torchtune builder functions themselves already return randomly-initialized models — the feature request is about surfacing that at the scaffold layer without touching recipe source.

## Implementation sketch

Three touch-points in cruijff_kit:

- **Recipes** — `tools/torchtune/custom_recipes/lora_finetune_single_device_stable.py` and `lora_finetune_distributed_stable.py`: guard the checkpoint load (`model.load_state_dict(...)` in the single-device recipe, `training.load_from_full_model_state_dict(...)` in the distributed one) behind a `cfg.get("random_init", False)` check. Log a clear line so it's visible in training output.
- **Scaffold** — `tools/torchtune/setup_finetune.py`: accept `random_init` and optional `seed` from `setup_finetune.yaml`, forward into the rendered `finetune.yaml`.
- **Template** — `tools/torchtune/templates/finetune_template.yaml`: when `random_init: true`, skip the `checkpointer` block entirely so users don't need a checkpoint directory on disk at all.

One subtlety: torchtune's `RMSNorm` uses the attribute name `scale` rather than `weight`, which matters if anyone tries to apply a custom init function based on PyTorch conventions. The default builder init is fine; just worth noting.

## Coverage note

Verify on at least one tied-embedding model (e.g. Llama 3.2 1B) and one untied (e.g. Llama 3.1 8B) — the tied case (`tie_word_embeddings: true`) is the easier regression to miss.

## Nice-to-haves

- A utility (in `utils/` or similar) that applies Llama-style `N(0, 0.02)` init after construction, for users who want the canonical pretraining distribution rather than torchtune's builder default.
- An example config under `experiments/` demonstrating the flag.
- A note in `WORKFLOW_GUIDE.md` on when a random-init training run is useful.

## Companion issue

Eval-side random-init support is tracked separately: **[eval_config.yaml companion issue #TBD]**. The two pair naturally — train from random init, then eval that checkpoint as an architecture-only baseline — but they can be implemented independently.

## Acceptance criteria

**Scaffold layer (must pass in PR CI):**

- [ ] `random_init: true` in `setup_finetune.yaml` flows through `setup_finetune.py` into the rendered `finetune.yaml` (new unit test, mirroring existing patterns in `tests/unit/test_setup_finetune.py`)
- [ ] With `random_init: true`, the rendered `finetune.yaml` has no `checkpointer:` block pointing at on-disk weights — a user with no checkpoint directory on disk can still run the training
- [ ] With `random_init: false` (default), rendered output is byte-identical to the current baseline (regression guard against the existing `tests/fixtures/scaffold/torchtune/rank{4,8}/` fixtures)
- [ ] New golden fixture `tests/fixtures/scaffold/torchtune/rank4_random_init/` demonstrating a minimal valid config
- [ ] `random_init: true` combined with `resume_from_checkpoint: true` raises a clear error (mutually exclusive)
- [ ] `seed` is optional; when provided, it's forwarded to the recipe and applied before model construction

**Recipe layer (must pass in PR CI):**

- [ ] Unit test: with `cfg.random_init == True`, the recipe's `_setup_model()` skips `model.load_state_dict(...)` (single-device) and `training.load_from_full_model_state_dict(...)` (distributed). Mock the checkpointer and assert it's not called.
- [ ] Clear log line emitted at training start so `random_init` runs are unmistakable in SLURM logs

**End-to-end (nightly GPU smoke test):**

- [ ] A ~50-step training run with `random_init: true` on Llama 3.2 1B completes without error and writes a checkpoint
- [ ] Initial loss is within [log(vocab_size) ± 0.5] (≈11.0–12.5 for a 128k vocab) — evidence the weights are genuinely random, not silently loaded from somewhere
- [ ] Same run with `seed: 42` produces byte-identical initial-step logits across two fresh invocations (determinism)
- [ ] Smoke test passes on one tied-embedding model (Llama 3.2 1B) and one untied (Llama 3.1 8B) — the tied case is the easier regression to miss

**Docs:**

- [ ] `WORKFLOW_GUIDE.md` gains a short section on when to use `random_init`
- [ ] At least one example config under `experiments/` uses the flag

---

## Background

This is motivated by research workflows where the architecture needs to be studied independently of the pretrained weights. A randomly-initialized model serves as a principled null baseline: same architecture, same tokenizer, same inference pipeline, only the weights differ. Running the same cruijff_kit pipeline against both a pretrained-and-fine-tuned model and an architecture-only model gives a clean measurement of what pretraining contributes on a given task, which is standard practice in the interpretability and architecture-analysis literature.

Underlying use cases:

- Computational social science work comparing pretrained-model behavior against an architecture-only null
- Interpretability research probing activation statistics at initialization
- Training runs from random init at non-standard scales (shorter runs for methodology development; ablations over architectural hyperparameters)
- Clean baselines for fine-tuning experiments — "what does fine-tuning actually add on top of random init?"

For training from scratch (as opposed to evaluating random init directly), note that the default learning-rate schedules in cruijff_kit's recipes are tuned for fine-tuning (~1e-5 to 1e-4). Training from scratch wants closer to 3e-4 with warmup — worth flagging in the example config if one is added.

**Comment by niznik-dev** [comment]
<img src="https://raw.githubusercontent.com/niznik-dev/llm-personas/main/fusions/mattie/mattie-icon.png" width="36" align="absmiddle"> 🎻 **MxC** ✳️

Closing as part of the project archive. The spec here is complete and the manual workaround still works. One caveat recorded in the handoff notes for whoever picks it up: with a LoRA recipe the random base weights stay frozen, so a meaningful architecture-only baseline likely needs full fine-tuning, which the custom recipes don't currently do.

—MxC
